src/java/org/apache/cassandra/index/sai/disk/v1/SegmentMetadataBuilder.java

```diff
@@ -0,0 +1,396 @@
+
+        @Override
+        public int getDocCount()
+        {
+            return values.getDocCount();
+        }
+
+        @Override
+        public long size()
+        {
+            return values.size();
+        }
+
+        @Override
+        public void getValue(int i, BytesRef packedValue)
+        {
+            values.getValue(i, packedValue);
+        }
+
+        @Override
+        public byte getByteAt(int i, int k)
+        {
+            return values.getByteAt(i, k);
+        }
+
+        @Override
+        public int getDocID(int i)
+        {
+            return values.getDocID(i);
+        }
+
+        @Override
+        public void swap(int i, int j)
+        {
+            values.swap(i, j);
+        }
+
+        @Override
+        public byte[] getMinPackedValue()
+        {
+            return values.getMinPackedValue();
+        }
+
+        @Override
+        public byte[] getMaxPackedValue()
+        {
+            return values.getMaxPackedValue();
+        }
+
+        @Override
+        public int getNumDimensions()
+        {
+            return values.getNumDimensions();
+        }
+
+        @Override
+        public int getBytesPerDimension()
+        {
+            return values.getBytesPerDimension();
+        }
+
+        @Override
+        public int getNumIndexDimensions() throws IOException
+        {
+            return values.getNumIndexDimensions();
+        }
+
+        @Override
+        public PointTree getPointTree() throws IOException
+        {
+            return values.getPointTree();
+        }
+
+        @Override
+        public void intersect(IntersectVisitor visitor) throws IOException
+        {
+            values.intersect((docId, term) -> {
+                if (!Arrays.equals(term, lastTerm))
+                {
+                    if (lastTerm != null)
+                        builder.add(encodeIfNeeded(lastTerm), count);
+
+
+                    count = 0;
+                    lastTerm = Arrays.copyOf(term, term.length);
+                }
+                count++;
+                visitor.visit(docId, term);
+            });
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            if (lastTerm != null)
+            {
+                builder.add(encodeIfNeeded(lastTerm), count);
+            }
+        }
+
+        /**
+         * Converts the term values coming from the index into a proper byte comparable values.
+         * This is needed because values of some types are stored in the index using their
+         * raw serializations, without encoding; but some others use byte comparable serialization.
+         */
+        private ByteComparable encodeIfNeeded(byte[] term)
+        {
+            AbstractType<?> type = builder.validator;
+            if (type instanceof InetAddressType || type instanceof IntegerType || type instanceof DecimalType)
+                return (v) -> type.asComparableBytes(ByteBuffer.wrap(term), v);
```

Comment:
What do you think about moving this logic to the `TypeUtil` class? SAI treats these three as special cases, and I think it could be helpful to keep that complexity in one place.

Reply:
+1